docs/RFCS/20190731_cluster_name.md

```diff
@@ -0,0 +1,173 @@
+  connect to multiple CockroachDB clusters simultaneously.
+
+Out of scope: it would be good to check the cluster name for equality
+in the node certificate properties. This would prevent a node from
+trusting another node's certificate if not in the same cluster, even
+though they may be signed by the same CA. See https://github.com/cockroachdb/cockroach/issues/28408
+
+
+# Motivation
+
+The motivation for doing this has been detail in
+https://github.com/cockroachdb/cockroach/issues/16784 and
+https://github.com/cockroachdb/cockroach/issues/15888
+
+- avoid mistaken node joins
+- disambiguate UI screens
+- disambiguate clusters in prometheus metrics
+
+# Guide-level explanation
+
+The service manager in charge of starting the `cockroach` process
+would activate this feature by providing the `--cluster-name` parameter
+on the command line.
+
+When the parameter is supplied, a newly created node will verify that
+the other nodes that it connects to (via `--join`) have the same name
+configured. If they detect a different name, the join would fail and the
+user would be informed.
+
+If the parameter is not supplied, we get the current behavior - the
+join succeeds in any case.
+
+There are two scenarios of interest:
+
+- newly-created clusters. These would have `--cluster-name` set on
+  every node from the beginning.
+
+- previously-created clusters that don't have `--cluster-name`
+  configured yet, and wish to "opt into" the system.
+  For these the following process is required:
+
+  1. restart (possibly with a version upgrade to 19.2) every node in a
+     rolling fashion, specifyfing both `--cluster-name`
+     `--disable-peer-name-verification` on the command line.
+  2. perform another rolling restart, removing `--disable-peer-name-verification`
+     from every node.
+
+# Reference-level explanation
+
+- The parametes `--cluster-name` and
+  `--disable-peer-name-verification` are parsed for `cockroach start`
+  only.
+
+- A maximum of 256 characters (for now) is allowed for `--cluster-name`.
+
+- Its lexical format is verified against the following regexp: `^[a-zA-Z](?:[-.a-ZA-Z0-9]*[a-zA-Z0-9]|)$`
+  (allows `a.b.c` and `a123` and `a-b` but not `a.`, `123a` or `b-`.
+  We choose to exclude "_" because we foresee this may want
+  to become a valid hostname eventually).
+
+- It is stored in the `server.Config` object, and propagated to the RPC
+  context and the heartbeat service.
+
+- The name is populated by the recipient of a PingRequest into the
+  PingResponse object, and checked by the initiator of the heartbeat. We choose
+  to check in the initiator so that the error condition can be
+  reported clearly to the operator (via `log.Shout`).
+
+- The flag `--disable-peer-name-verification` alters the verification as follows:
+  when set, if *either* side has an empty cluster name, the comparison is skipped.
+  In the case where both side have a name configured, the comparison is still
```

Comment:
What's the reasoning behind doing the validation if both are set? It seems like a bit of a surprise, and does not help the scenario where one really wants to change the name (you would need to transition through an empty cluster name, then switch to the new name).

Reply:
Nevermind, keeping the name consistent is a pretty good reason. I'm wondering if we need to worry about the "change name" case. It should be rare and is still feasible.

Reply:
sorry by "either" here I meant "either/or". 
If we want to facilitate the cluster name change later while a cluster is up, we can invent a cluster RPC that sets the "disable peer name verification" flag on all running nodes, then changes the name, then sets the flag back. I'll write it in the RFC.